### PR TITLE
PR: Move LSP related code from the `CodeEditor` class definition to a new mixin and `CodeEditor` related elements to a `codeeditor` module (Editor)

### PR DESCRIPTION
**Pull request by dalthviz**
## Description of Changes

Reduce `CodeEditor` class definition size and improve Editor widgets organization by moving LSP related code from the `CodeEditor` class definition to a new mixin class in a different file. Also, move the `TabSwitcherWidget` class to its own file and create a `codeeditor` directory for code and test related with the `CodeEditor` class 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: dalthviz

### CODE REVIEW

File: spyder/plugins/editor/widgets/lsp.py
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/lsp.py:1]: I suggest ti rename this file to `lsp_mixin.py`.

**Review by ccordoba12 — COMMENTED** [review]
@dalthviz, this looks good to me and I think it's fine for now. But what do you think about creating a `codeeditor` module in `spyder/plugins/editor/widgets` and move there `codeeditor.py`, `lsp_mixin.py` and all tests that only check the `CodeEditor` class?

Furthermore, I think we should do the same for `editorstack.py` too.

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 1]
```diff
@@ -0,0 +1,1330 @@
...
+    class_register,
+)
+from spyder.plugins.editor.panels import FoldingPanel
+from spyder.plugins.editor.panels.utils import (
+    merge_folding,
+    collect_folding_regions,
+)
+from spyder.plugins.editor.utils.editor import BlockUserData
+from spyder.utils import sourcecode
+
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:49]:
```suggestion


```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 2]
```diff
@@ -0,0 +1,1330 @@
...
+            self._document_server_needs_update = False
+
+        # Send pending requests
+        for method, params, requires_response in self._pending_server_requests:
+            self.emit_request(method, params, requires_response)
+
+        # Clear pending requests
+        self._pending_server_requests = []
+
+    # ---- LSP: Basic methods
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:215]:
```suggestion
    # ---- Basic methods
```

We don't need to mention LSP here.

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 3]
```diff
@@ -0,0 +1,1330 @@
...
+            logger.error(message, exc_info=True)
+        else:
+            # We need this because logger.error activates our error
+            # report dialog but it doesn't show the entire traceback
+            # there. So we intentionally leave an error in this call
+            # to get the entire stack info generated by it, which
+            # gives the info we need from users.
+            logger.error("%", 1, stack_info=True)
+
+    # ---- LSP: Configuration and start/stop
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:252]:
```suggestion
    # ---- Configuration and start/stop
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 4]
```diff
@@ -0,0 +1,1330 @@
...
+            "version": self.text_version,
+            "text": text,
+            "codeeditor": self,
+            "offset": cursor.position(),
+            "selection_start": cursor.selectionStart(),
+            "selection_end": cursor.selectionEnd(),
+        }
+        return params
+
+    # ---- LSP: Symbols
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:379]:
```suggestion
    # ---- Symbols
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 5]
```diff
@@ -0,0 +1,1330 @@
...
+        except RuntimeError:
+            # This is triggered when a codeeditor instance was removed
+            # before the response can be processed.
+            return
+        except Exception:
+            self.log_lsp_handle_errors("Error when processing symbols")
+        finally:
+            self.symbols_in_sync = True
+
+    # ---- LSP: Linting and didChange
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:414]:
```suggestion
    # ---- Linting and didChange
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 6]
```diff
@@ -0,0 +1,1330 @@
...
+                # Don't append messages to data for cloned editors to avoid
+                # showing them twice or more times on hover.
+                # Fixes spyder-ide/spyder#15618
+                if not self.is_cloned:
+                    data.code_analysis.append(
+                        (source, code, severity, message)
+                    )
+                block.setUserData(data)
+
+    # ---- LSP: Completion
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:646]:
```suggestion
    # ---- Completion
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 7]
```diff
@@ -0,0 +1,1330 @@
...
+        except RuntimeError:
+            # This is triggered when a codeeditor instance was removed
+            # before the response can be processed.
+            return
+        except Exception:
+            self.log_lsp_handle_errors(
+                "Error when handling completion item resolution"
+            )
+
+    # ---- LSP: Signature Hints
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:794]:
```suggestion
    # ---- Signature Hints
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 8]
```diff
@@ -0,0 +1,1330 @@
...
+                    documentation=documentation,
+                )
+        except RuntimeError:
+            # This is triggered when a codeeditor instance was removed
+            # before the response can be processed.
+            return
+        except Exception:
+            self.log_lsp_handle_errors("Error when processing signature")
+
+    # ---- LSP: Hover/Cursor
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:853]:
```suggestion
    # ---- Hover/Cursor
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 9]
```diff
@@ -0,0 +1,1330 @@
...
+                )
+                self._last_point = None
+        except RuntimeError:
+            # This is triggered when a codeeditor instance was removed
+            # before the response can be processed.
+            return
+        except Exception:
+            self.log_lsp_handle_errors("Error when processing hover")
+
+    # ---- LSP: Go To Definition
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:918]:
```suggestion
    # ---- Go To Definition
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 10]
```diff
@@ -0,0 +1,1330 @@
...
+        except RuntimeError:
+            # This is triggered when a codeeditor instance was removed
+            # before the response can be processed.
+            return
+        except Exception:
+            self.log_lsp_handle_errors(
+                "Error when processing go to definition"
+            )
+
+    # ---- LSP: Document/Selection formatting
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:966]:
```suggestion
    # ---- Document/Selection formatting
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 11]
```diff
@@ -0,0 +1,1330 @@
...
+                cursor = self.textCursor()
+                cursor.movePosition(
+                    QTextCursor.Down,
+                    QTextCursor.MoveAnchor,
+                    self.__line_number_before_format,
+                )
+                self.setTextCursor(cursor)
+                self.centerCursor()
+
+    # ---- LSP: Code folding
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:1194]:
```suggestion
    # ---- Code folding
```

File: spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py [hunk 12]
```diff
@@ -0,0 +1,1330 @@
...
+        # Fixes spyder-ide/spyder#19514
+        if self._folding_info is not None:
+            folding_panel.update_folding(self._folding_info)
+
+        # Update indent guides, which depend on folding
+        if self.indent_guides._enabled and len(self.patch) > 0:
+            line, column = self.get_cursor_line_column()
+            self.update_whitespace_count(line, column)
+
+    # ---- LSP: Save/close file
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/lsp_mixin.py:1290]:
```suggestion
    # ---- Save/close file
```

File: spyder/plugins/editor/widgets/codeeditor/tests/conftest.py [hunk 13]
```diff
@@ -0,0 +1,316 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+#
+
+"""
+Testing utilities for the CodeEditor to be used with pytest.
+"""
+# Stdlib imports
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/tests/conftest.py:10]:
```suggestion

# Standard library imports
```

**Review by ccordoba12 — COMMENTED** [review]
Great refactoring @dalthviz! I left some minor comments, the rest looks good to me.

*Note*: I think you mostly moved code around, so I wasn't so thorough in my review. But let me know if you want me to pay attention to something specific.

**Comment by dalthviz** [comment]
@ccordoba12 maybe something to give a more in depth check are the `conftest.py` definitions and the assets usage and location. Also, while checking some tests related with autoformatting locally, they were failing for me. So maybe those failing tests could be related with #20898

File: spyder/plugins/editor/widgets/codeeditor/tests/conftest.py [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+    qtbot_module
+)
+from spyder.plugins.editor.widgets.codeeditor import CodeEditor
+from spyder.plugins.explorer.widgets.tests.conftest import create_folders_files
+from spyder.plugins.outlineexplorer.editor import OutlineExplorerProxyEditor
+from spyder.plugins.outlineexplorer.main_widget import OutlineExplorerWidget
+from spyder.utils.programs import is_module_installed
+from spyder.widgets.findreplace import FindReplace
+
+# ---- Base constants for assets
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/tests/conftest.py:37]:
```suggestion

# ---- Base constants for assets
```

File: spyder/plugins/editor/widgets/codeeditor/tests/conftest.py [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+
+    # Fix the size of the outline explorer to prevent an
+    # 'Unable to set geometry ' warning if the test fails.
+    outlineexplorer.setFixedSize(400, 350)
+
+    qtbot.addWidget(outlineexplorer)
+    outlineexplorer.show()
+
+    return outlineexplorer
+
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor/tests/conftest.py:119]:
```suggestion


```

**Review by ccordoba12 — COMMENTED** [review]
> @ccordoba12 maybe something to give a more in depth check are the conftest.py definitions and the assets usage and location.

That looks good to me. I just left a couple of formatting comments on `conftest.py`.

> Also, while checking some tests related with formatting/autoformatting locally, they were failing for me. Maybe those failing tests could be related with https://github.com/spyder-ide/spyder/issues/20898

I ran `test_formatting.py` locally using the latest Black version and all tests there are passing for me

> Edit: Also, some tests related with the outline seems like fail locally

I also ran `test_outline.py` and it passed for me.

I wouldn't worry too much about it because sometimes tests also fail for me locally.

**Pull request merged**